Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/Specifications/SqlTableValuedParameterComparedColumnSpecification.cs

```diff
@@ -0,0 +1,111 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Text;
+using Microsoft.SqlServer.Server;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders.Specifications;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.DbCommandBuilders.Specifications;
+
+/// <summary>
+/// Specifies a set of <see cref="ObjectValues"/> that are transmitted as a table-valued parameter, to be used in a WHERE-IN-clause.
+/// </summary>
+public class SqlTableValuedParameterComparedColumnSpecification : IComparedColumnsSpecification
+{
+  /// <summary>
+  /// The column whose values are looked up in the <see cref="ObjectValues"/>.
+  /// </summary>
+  public ColumnDefinition ColumnDefinition { get; }
+
+  /// <summary>
+  /// The values that fulfill the filter criterion.  
+  /// </summary>
+  public IEnumerable<object?> ObjectValues { get; }
+
+  public SqlTableValuedParameterComparedColumnSpecification (ColumnDefinition columnDefinition, IEnumerable<object?> objectValues)
+  {
+    ArgumentUtility.CheckNotNull(nameof(columnDefinition), columnDefinition);
+    ArgumentUtility.CheckNotNull(nameof(objectValues), objectValues);
+
+    ColumnDefinition = columnDefinition;
+    ObjectValues = objectValues;
+  }
+
+  /// <inheritdoc />
+  public void AddParameters (IDbCommand command, ISqlDialect sqlDialect)
+  {
+    ArgumentUtility.CheckNotNull(nameof(command), command);
+    ArgumentUtility.CheckNotNull(nameof(sqlDialect), sqlDialect);
+
+    var storageTypeInfo = ColumnDefinition.StorageTypeInfo;
+
+    var parameterName = GetParameterName(sqlDialect);
+    var parameter = (SqlParameter)sqlDialect.CreateDataParameter(command, storageTypeInfo, parameterName, null);
+
+    // use the parameter we need anyway to find out the SqlDbType for the column
+    parameter.DbType = storageTypeInfo.StorageDbType;
+    var sqlMetaData = new SqlMetaData("Value", parameter.SqlDbType);
+
+    var typeName = $"TVP_{storageTypeInfo.StorageDbType}";
+
+    parameter.SqlDbType = SqlDbType.Structured;
+    parameter.TypeName = typeName;
+
+    if (ObjectValues.Any())
+    {
+      var tvpValue = new SqlTableValuedParameterValue(typeName, new[] { sqlMetaData });
+      foreach (var value in ObjectValues)
+      {
+        var convertedValue = storageTypeInfo.ConvertToStorageType(value);
+        tvpValue.AddRecord(convertedValue);
+      }
+
+      parameter.Value = tvpValue;
+    }
+    else
+    {
+      parameter.Value = DBNull.Value;
+    }
```

Comment:
didn't I see this logic already somewhere? in particular, empty-tvp is dbnull

Reply:
Yes, and it's wrong here - the value should ne `null`, not `DBNull.Value` - we can delete the `else`-branch.
The other place is in `SqlTableValuedDataParameterDefinition`.

Reply:
and the rest of the copy/convert logic? is this also implemented in SqlTable..ParamDef? What would be required to unify the logic, assuming that it is not just accidentally the same logic but the same business case.

Reply:
We could try to put the conversion into `SqlTableValuedParameterValue`, but that would require injecting the conversion delegates for each column along with the `SqlMetaData`, using either a wrapper type or `ValueTuple` as item type for the `IReadOnlyCollection<T>`.
This would mean eliminating the conversion from `RecordDefinition.GetColumnValues()`.
I can do that in a separate commit, so you can see what it would look like.